### ISSUE TITLE
3.2 questions and comments

### ISSUE THREAD

**Issue by nicm** [tmux/tmux#2645]
### Ask or discuss any quick questions here about 3.2 please.

Bug reports please open a new issue with full information, see [CONTRIBUTING](https://github.com/tmux/tmux/blob/master/.github/CONTRIBUTING.md).

**[The 3.2 CHANGES file is here](https://raw.githubusercontent.com/tmux/tmux/3.2/CHANGES).**

As always, thanks to the many people who contributed to both this and to previous releases.

---

There are a **lot** of changes in this release, these are the most notable, for more see the CHANGES file:

- How data is sent to control mode clients has been completely revamped to both be more fair with multiple panes and to prevent huge amounts of data being backed up; this feature is used by default when tmux is attached with the latest iTerm2 nightly builds.

- Configuration file parsing has changed slightly: the contents of the new {} syntax introduced in 3.1 must now be valid tmux command syntax; and to allow formats to be annotated, strings given with quotes may now contain newlines (leading spaces and comments are stripped).

- A new customize mode available with `C-b C` (`C-b S-c`) which allows options and key bindings to be browsed and changed interactively.

- Support for extended keys offered by some terminals (xterm, mintty, iTerm2), see https://github.com/tmux/tmux/wiki/Modifier-Keys#extended-keys.

- A `pane-border-lines` option to change the characters used to draw the pane border separators.

- How UTF-8 data is stored has been rewritten to reduce memory use for characters in the BMP.

- The message log (`C-b ~`) has been changed to be per server instead of per client and to have some useful content.

- Client flags have been extended (`refresh-client -F` becomes `-f` and also added for `attach-session` and `new-session`), and a `read-only` flag added (same as `-r` to `attach-session`).

- A new `active-pane` client flag that if given allows a client to have its own active pane for each window rather than being tied to the server's active pane.

- Improved `command-prompt` completion including showing a menu of completions.

- All style options can now be formats, for example the default `pane-active-border-style` now changes colour depending on `pane_in_mode` and `synchronize-panes`.

- Search match count has come back in copy mode, as well as performance improvements and additional styles for marking of search terms.

- Try `$XDG_CONFIG_HOME/tmux/tmux.conf` as well as `~/.config/tmux/tmux.conf`.

- Window and pane hooks such as `window-layout-changed` and `pane-exited` are now window or pane options instead of session options.

- `copy-pipe` and friends can now be given with no arguments in which case they get the command from a new `copy-command` option. The default copy commands all now use `copy-pipe`. This makes it much easier to configure `copy-pipe` for all copy key bindings.

- Terminal detection and feature code has been rewritten. tmux now collects sets of related terminfo(5) capabilties together with sensible default values into named "terminal features" which makes them easier to configure via a new `terminal-features` option or `-T` flag to tmux itself.  tmux can detect and will automatically configure a number of common, modern terminal emulators.

- Support for per-client overlay popups with a new `display-popup` command. A popup is also used for an `e` key binding in buffer mode to edit a buffer using the editor given in the new `editor` option (filled in from the `EDITOR` or `VISUAL` environment variables by default).

- `DoubleClick` and `TripleClick` mouse key bindings have changed so that only one is fired. Behaviour similar to the old double-click key binding is now available as `SecondClick`. There are now default double- and triple-click bindings to copy a word or line, and a default middle-click binding to paste the top buffer.

- When a server socket is given with `-S`, it is now created with umask 177 rather than 117.

---

**Note the following known potentially-incompatible changes:**

- The {} syntax added in tmux 3.1 has changed so that the contents must follow the syntax of tmux commands.

- The `DoubleClick` key binding is not fired on the way to `TripleClick`; only one is fired. The new `SecondClick` key binding behaves similarly to the old `DoubleClick`.

- `status-fg` and `status-bg` are now applied before `status-style` rather than instead of; this means that if either is set to `default` the `status-style` colour will be used. To fix, replace `status-bg default` with `status-style bg=default`.

**Comment by hanh090** [comment]
I want to ask a question related to `pane-title` length, if I have a long command seems like it was truncated to 80 chars only, I am using version 3.1c, is that fix in latest version?

**Comment by nicm** [comment]
tmux should not truncate it, how are you setting it?

**Comment by hanh090** [comment]
@nicm my setup is here: https://github.com/hanh090/dotfiles/blob/master/.tmux.conf when I run it in REPL, it always truncate for lower 80 characters

**Comment by nicm** [comment]
Sounds like a problem with your shell then.

**Comment by hanh090** [comment]
> Sounds like a problem with your shell then.

![g1XgqVscSn](https://user-images.githubusercontent.com/3643657/114531408-122dd680-9c76-11eb-994d-3cddb70bd899.gif)
this is sample, can you try it?

**Comment by nicm** [comment]
I don't know what you are asking. If it works when you wrap it in a shell function it must be a shell problem not a tmux one.

**Comment by pschmitt** [comment]
Congrats on the new release!

I'm currently trying to build a [termux package](https://github.com/termux/termux-packages/blob/master/packages/tmux/build.sh) for it but I'm not able to go past the `./configure` step:

```shell
[...]
checking for strndup... yes
checking for strsep... yes
checking for strtonum... no
checking for working strnlen... yes
checking for working reallocarray... configure: error: in `/home/builder/.termux-build/tmux/src':
configure: error: cannot run test program while cross compiling
See `config.log' for more details
```

Can someone give me a hint? Cross-compiling is hard :)

**Comment by nicm** [comment]
Did this work before? ISTR someone complaining that newer autoconf cannot run some tests while crosscompiling. You may need to patch configure.ac to hardcode this test for your platform and regenerate configure, or just patch configure.

**Comment by pschmitt** [comment]
For 3.1c it worked fine, yes.

I'll try to craft a quick patch to disable the check, even though I really don't if it is safe to assume that reallocarray is available (or not?) on Android.

**Comment by nicm** [comment]
Not sure if it has it or not but if you assume it is not present it should be OK.

**Comment by pschmitt** [comment]
Thanks for your help. I've submitted my PR to [termux-packages](https://github.com/termux/termux-packages/pull/6625).

<details>
  <summary>🥳 Greetings from Termux!</summary>

![image](https://user-images.githubusercontent.com/37886/114543225-3202d180-9c59-11eb-9549-099f23e06e41.png)
</details>

**Comment by bertin0** [comment]
Just got this update on Arch, and it seems to have broken my user-key binding for using Ctrl+; to switch panes (i use jkl; instead of hjkl):

```
set -s user-keys[0] "\e[27;5;59~"
bind User0 next-window
```

**Comment by nicm** [comment]
What happens if you bind `C-+` instead?

**Comment by bertin0** [comment]
Unsure how i would test that, as both of my terminals (alacritty ant termite) use it to zoom in and I don't know what escape sequence it would have. I tested `C-'`, `C-.`, `C-,` and none worked. (for the escape sequences generated by termite).

**Comment by nicm** [comment]
Oh  sorry I meant `C-;`.

**Comment by bertin0** [comment]
Nope, `C-;` or `C-\;` didn't work for me

**Comment by nicm** [comment]
Does this fix it?
~~~~
--- tty-keys.c  13 Apr 2021 05:23:34 -0000      1.145
+++ tty-keys.c  13 Apr 2021 11:36:30 -0000
@@ -969,6 +969,8 @@ tty_keys_extended_key(struct tty *tty, c
                                onlykey = 0;
                        else if (onlykey == 63)
                                onlykey = 127;
+                       else
+                               onlykey |= KEYC_CTRL;
                        onlykey |= ((nkey & KEYC_MASK_MODIFIERS) & ~KEYC_CTRL);
                }
                nkey = onlykey;

**Comment by nicm** [comment]
Actually please try this instead: https://gist.githubusercontent.com/nicm/be4faa3ff53884384695fc271b3428fe/raw/6cfa7f662a61eae060e8878c6230dfceb6879c70/gistfile1.txt

**Comment by bertin0** [comment]
Unfortunately not. Without that patch, tmux woutld just output `;` on `C-;`, with this patch it seems to not do anything.  (both for `User0` and `C-\;`/`C-;`

**Comment by nicm** [comment]
It works  for me, open a new issue with logs please.

**Comment by boris-petrov** [comment]
@nicm - thanks for the new release! I want to go back to [this discussion](https://github.com/tmux/tmux/issues/2186#issuecomment-623108894) we had. 3.2 shows correctly the matched count when searching, however it doesn't show which match I am on while going back through them with `n`. For example, tmux 3.0 showed `(1/3 results)` when I search for something, then `2/3` on the next result and so on. tmux 3.2 shows only `(3 results)` with no indication which result I'm on right now.

Is it possible to bring back the old behavior?

**Comment by bertin0** [comment]
> It works for me, open a new issue with logs please.

Sorry, I must have not reloaded my conf properly, it works now! Thanks for the quick fix!

**Comment by nicm** [comment]
OK I have applied this and it will be in 3.2a but I will wait for a while to see if there is anything else.

**Comment by cah-justin-pitts02** [comment]
I encountered a parsing issue in .tmux.conf and I fixed it but could not find any documentation related directly to the issue. 

This:
    `bind-key m send-keys -t 0.{top-left} n`
used to work. 

Now I get a syntax error on this line on startup. 

If i quote the argument to -t like so
    `bind-key m {send-keys -t ':0.{top-left}' n}`
It does work. 

Looking at the server log, it appears the lexer was not getting the close brace for top-left.

**Comment by sschlesier** [comment]
I have a similar issue to @cah-justin-pitts02

I've been using the below for some time, in 3.2 it results in a syntax error.

```
# Create new pane in current directory
bind -n M-n split-window -b -t :.{top-left} -c "#{pane_current_path}"
```

I've not yet puzzled out how to fix this line.  I assume it has something to do with the breaking changes to the `{}` syntax mentioned above.  I did not find any other documentation about what those changes are.

**Comment by cah-justin-pitts02** [comment]
@sschlesier strange - that works unmodified for me. If I were to apply my fix to your line, it would be

`bind -n M-n split-window -b -t ':.{top-left}' -c "#{pane_current_path}"`

**Comment by JonRowe** [comment]
Did something change for named sessions? `tmux new -s <name>` now immediatly exits and deposits a rubbish string from item
![Screen Shot 2021-04-13 at 19 22 27](https://user-images.githubusercontent.com/162976/114601604-98a1e280-9c8d-11eb-8eab-9d7344c4b7e7.png)

**Comment by nicm** [comment]
@JonRowe most likely your shell or `default-command` doesn't work so tmux immediately exits.

**Comment by Rylan12** [comment]
After upgrading to tmux 3.2 my `M-Up` and `M-Down` keybindings have stopped working. I'm using iTerm2 on macOS (with tmux installed from Homebrew)

In `~/.tmux.conf`:

```tmux
bind-key -n M-Up next-window
bind-key -n M-Down previous-window
```

I glanced at the changelog and saw nothing indicating a change with these. Is there an obvious fix or should I open a separate issue?

**Comment by nicm** [comment]
What do you see if you run `cat` outside tmux and press `M-Up`?

**Comment by Rylan12** [comment]
In iTerm2 (both in and out of tmux):
```console
$ cat
^[[1;9A
```

In the default terminal app (although `M-Up` still doesn't work properly):
```console
$ cat
^[[A
```

**Comment by nicm** [comment]
`^[[1;9A` is not `M-Up` - `M-Up` would be `^[[1;3A` or `^[^[A`. tmux has never recognised 9 as Meta so I don't know why this would have ever worked. What version did you upgrade from? Have you also upgraded iTerm2 recently?

**Comment by Rylan12** [comment]
I updated tmux to version 3.2 from 3.1. I haven't manually updated iTerm2 recently (or seen a popup letting me know that it's updating) but I am running the current version.

Would a solution be to map `M-Up` to `^[[1;3A` in the iTerm2 profile?

(Heading home so I may no be able to respond for a half-hour or so)

**Comment by nicm** [comment]
Yes that will fix it, but you are the second person to ask about this today so I wonder what has changed, because tmux has never handled `^[[1;9A` as `M-Up` and I do not see anything obvious in recent iTerm2 changes. My iTerm2 does not have this configured but I installed it a year or so ago and have no idea if I changed the options (although it seems unlikely I would have configured the specific key escape sequences):
![Screenshot 2021-04-13 at 20 21 23](https://user-images.githubusercontent.com/304243/114608679-d60a6e00-9c95-11eb-855e-7156f789d0ef.png)

**Comment by Rylan12** [comment]
> My iTerm2 does not have this configured but I installed it a year or so ago and have no idea if I changed the options (although it seems unlikely I would have configured the specific key escape sequences):

Same here. I have done some configuration with keybindings in iTerm2 but not recently (and never with `M-Up` or `M-Down`). Adding those mappings now solved the problem.

I'd be happy to provide any other information that might help if you are going to pursue this further, so feel free to let me know if there's anything that would be helpful. My issue is solved, though, so don't feel any pressure from me. Thanks for your help!

**Comment by toxeus** [comment]
@nicm looks like this release broke `capture-pane`.
```sh
# hangs forever since 3.2
tmux capture-pane -J -p | cat
```

**Comment by sschlesier** [comment]
poking some more at the syntax error I reported earlier in the thread

If the command is sourced from a file it has a syntax error, but if typed as a command it works just fine.

I'll just go ahead and open a new issue with repro steps for this

**Comment by Frederick888** [comment]
Clipboard integration under macOS is broken for me from 3.2.

Steps to reproduce (tmux started without any user configuration):

1. `tmux set-window-option -g mode-keys vi`
0. `tmux copy-mode`
0. Select and copy some text
0. Repeat 1 & 2 for a few times
0. Command + v

Observed result: It seems that tmux keeps appending the selected text to clipboard rather than replacing its contents. `tmux save-buffer -` output is correct though.

**Comment by nicm** [comment]
What terminal are you using?

**Comment by boris-petrov** [comment]
@nicm - sorry for the spam, but did you see [my comment](https://github.com/tmux/tmux/issues/2645#issuecomment-818684547)?

**Comment by nicm** [comment]
@boris-petrov I don't think we want to add this back because it means searching the history to find all the matches which is very slow.

**Comment by boris-petrov** [comment]
@nicm - no, I mean to show which result is selected of the ones that have been found. The search is limited to 200ms which is totally fine. Then, when I go through the results, I would like to have the index of the current one *of the found ones*. Not of the "true" results count. This is good enough and is better than what is right now in 3.2. For the very least because it's hard to see when you "wrap" through the results and you jump from the last to the first.

**Comment by nicm** [comment]
It was removed because it didn't work in 3.1, it is much faster not to try and keep track of it, and neither vi or emacs show a match count. I don't think we need it.

**Comment by nicm** [comment]
I think it would be better to add some indicator when you wrap.

**Comment by boris-petrov** [comment]
> neither vi or emacs show a match count

Not in the default, no. I use `vim-airline` and it shows both the total count and the current result. By the way, we already had a [similar discussion](https://github.com/tmux/tmux/issues/2186#issuecomment-623114547). :smile: 

And again - right now I'm not talking about showing the total amount of results. The way it works in 3.2 is totally fine. I just would like showing the *current* result index. I don't see how that would impact performance at all? Shouldn't it be just a simple counter?

> I think it would be better to add some indicator when you wrap.

That would be nice but I don't think it replaces what I request.

**Comment by Frederick888** [comment]
> What terminal are you using?

I'm using Kitty and I just noticed that with tmux 3.1c this happens as well.

What has changed though is that, previously enabling [oh my tmux](https://github.com/gpakosz/.tmux) somehow made this problem disappear; now with or without it, this issue happens regardless.

It doesn't happen when using iTerm btw.

**Comment by nicm** [comment]
@Frederick888 this is a known bug in Kitty where it appends to the selection by default, I believe there is a way to fix it but you will need to talk to the Kitty developers.

**Comment by nicm** [comment]
@boris-petrov the only use case you mentioned is to show when you wrap so an indicator would fix that, no? What use is the current match?

Maybe we could do it as a counter where we just add and remove one every time you search again, the problem is that the way we were doing it meant we had to actually search everything again which is a pain.

**Comment by boris-petrov** [comment]
@nicm - that's one use case, yes. Another would be to just know where in the total results am I. For example, I run my app, I search for "Exception" in the stdout, I have 90 results, I start going through them and it is nice to know how many I have passed, how many are left, etc. This is just on the top of my head, I'm sure there are cases where it is just useful to know how far up the results you've gone.

I don't know your internal implementation, of course. I just hope it should be as simple as a counter. Because I really like and use that feature and I'll be sticking to 3.0 until I have it back. :smile:

**Comment by nicm** [comment]
@topcat001 what do you think about search result number? Can we just work out the initial result at the start and then plus/minus one every time we search again? Would that work? I don't want to force it to actually search everything again just to update a counter.

**Comment by Frederick888** [comment]
> @Frederick888 this is a known bug in Kitty where it appends to the selection by default, I believe there is a way to fix it but you will need to talk to the Kitty developers.

Thank you!

In case anyone's also interested, the Kitty configuration is https://sw.kovidgoyal.net/kitty/conf.html?highlight=clipboard#opt-kitty.clipboard_control

tl;dr Kitty users need the `no-append` option, e.g.

```
clipboard_control                       write-clipboard write-primary no-append
```

Previously it didn't happen because OMT overrides some bindings to use `pbcopy` directly. Why it stopped working is another problem though (but likely an OMT problem).

**Comment by topcat001** [comment]
> @topcat001 what do you think about search result number? Can we just work out the initial result at the start and then plus/minus one every time we search again? Would that work? I don't want to force it to actually search everything again just to update a counter.

It's a little tricky because for example what happens if the user moves the cursor past one or more results manually and then searches again? The counter should skip the results which were manually skipped. I'll spend some time on this tomorrow to see if there is a quick way to do this.

**Comment by LoLei** [comment]
Are there any possible breaking changes regarding [tpm](https://github.com/tmux-plugins/tpm) and/or [tmux-resurrect](https://github.com/tmux-plugins/tmux-resurrect)?

Activating tpm's install script or trying to save the state with tmux-resurrect freezes the current session after yesterday's tmux update. At least I think it lines up with that. I can provide more information (and here (https://github.com/tmux-plugins/tmux-resurrect/issues/379#issuecomment-819480891) is a comment where I describe it as well in further detail.)

**Comment by Zelnox** [comment]
I just updated and am unable to start a new tmux session on OSX 10.15.7 with iterm2 3.4.4.
Just starting `tmux -v` with no config, I get
```sh
tmux -v
[exited]
^[[?62;4c^[[ITERM2 3.4.4n
```
There are no tmux-server logs.

I can attach to a session created in the previous version of tmux.

**Comment by gpanders** [comment]
Hi @nicm, thanks for the new release!

Two questions:

1. I love that copy commands now copy the current search match if there is no selection. However, this doesn't seem to work for `pipe` commands. E.g. if I use `copy-pipe-and-cancel` in a binding without a selection, the search match is copied but it does not get piped to the command. Is this intentional?
2. I'd love some clarification on the new `terminal-features` setting. I read the [FAQ](https://github.com/tmux/tmux/wiki/FAQ#how-do-i-use-rgb-colour) and I added the following to `.tmux.conf`:
        
        set -as terminal-features ',*:RGB'

    However, `tmux info` still shows that RGB is `[missing]`. If I use `terminal-overrides` instead of `terminal-features`, then RGB reports `(flag) true` as expected. Does this mean that using `terminal-features` is not actually enabling RGB colors correctly, or is it just how it's being reported? (if it's relevant, I'm using iTerm2 and `TERM=screen-256color` in tmux).

**Comment by deviantfero** [comment]
The colors in my terminals are very different with this new version of tmux, see picture 1 with just this setting
```
set -g default-terminal screen-256color
```
![1618413172](https://user-images.githubusercontent.com/11150153/114734792-c759a080-9d01-11eb-98f1-e220d1cf43cb.png)

I previously fixed that by adding a `terminal-override` to tmux's settings, like so:
```
set -g default-terminal "screen-256color"
set -ga terminal-overrides ",*256col*:Tc"
```
but this yields even weirder colors now:
![1618413203](https://user-images.githubusercontent.com/11150153/114736094-089e8000-9d03-11eb-9376-ee4c3fab03cd.png)

I'm using `urxvt` as my terminal

**Comment by topcat001** [comment]
This is just a quick thought but we could search fully and enable the counter every time only if a timeout was not triggered. If the search timed out then then we could search only the visible area and drop the counter, since it does not make sense to maintain the counter in the latter case. This might be an acceptable compromise.

**Comment by ChrispyChris** [comment]
> The colors in my terminals are very different with this new version of tmux, see picture 1 with just this setting
> 
> ```
> set -g default-terminal screen-256color
> ```
> 
> ![1618413172](https://user-images.githubusercontent.com/11150153/114734792-c759a080-9d01-11eb-98f1-e220d1cf43cb.png)
> 
> I previously fixed that by adding a `terminal-override` to tmux's settings, like so:
> 
> ```
> set -g default-terminal "screen-256color"
> set -ga terminal-overrides ",*256col*:Tc"
> ```
> 
> but this yields even weirder colors now:
> ![1618413203](https://user-images.githubusercontent.com/11150153/114736094-089e8000-9d03-11eb-9376-ee4c3fab03cd.png)
> 
> I'm using `urxvt` as my terminal
> 
> **UPDATE:**
> found a consistent way to reproduce, download the following into a file called `sequences`: https://u.teknik.io/XXOWm
> 
> I tried pasting the contents here but it lost information due to it using special characters or something, posting the content just for reference but for it to work you're gonna need to use the file provided:
> 
> ```
> �]4;0;#1e0528�\�]4;1;#00d9e9�\�]4;2;#05cb0d�\�]4;3;#955ae7�\�]4;4;#550068�\�]4;5;#8991bb�\�]4;6;#b900b1�\�]4;7;#ffeeff�\�]4;8;#320f55�\�]4;9;#00d9e9�\�]4;10;#05cb0d�\�]4;11;#955ae7�\�]4;12;#550068�\�]4;13;#8991bb�\�]4;14;#b900b1�\�]4;15;#f8c0ff�\�]10;#ffeeff�\�]11;#1e0528�\�]12;#ffeeff�\�]13;#ffeeff�\�]17;#ffeeff�\�]19;#1e0528�\�]4;232;#1e0528�\�]4;256;#ffeeff�\�]708;#1e0528�\
> ```
> 
> then when inside tmux do `(cat /path/to/sequences &)`, do it also outside of tmux and see the difference, there might be some kind of error in interpreting the escape sequences that set colors in terminals.
> 
> can confirm this does not happen in version `3.1c`

I too am having odd colors now. The background now when I open tmux has a grey background instead of URXVT's transparent background. My laptop is still running 3.1c and the background works as transparent there. I then tried adding some different things:
set -g default-terminal "tmux-256color"
set -ga terminal-overrides ",*256col*:Tc"
set -ga terminal-overrides '*:Ss=\E[%p1%d q:Se=\E[ q'
set-environment -g COLORTERM "truecolor"

That fixed my transparent background but then completely broke my vim colors. Is there a way to get the old colors working like they were in 3.1c? I tried digging around the notes and am unsure what change broke this all. Here is a picture of my terminal when I open tmux now.
https://imgur.com/a/fWdDod0

**Comment by nicm** [comment]
@ChrispyChris open an issue with logs please.

@topcat001 that could work, I don't really mind so long as it isn't overly complicated and doesn't end up doing a full search all the time.

**Comment by nicm** [comment]
@deviantfero this file remaps the palette for the pane. If your terminal does not support RGB colour, or you have not configured it, then tmux will map each palette entry to the nearest colour in the standard 256-colour palette. If you want to use palette-setting and have accurate colours, you will need to either set the palette before starting tmux, or use a terminal which supports RGB colour and enable it (see https://github.com/tmux/tmux/wiki/FAQ#how-do-i-use-rgb-colour).

**Comment by deviantfero** [comment]
@nicm just tried with both those options, not working on 3.2
![1618497387](https://user-images.githubusercontent.com/11150153/114887459-b3777280-9dc5-11eb-9698-6bc2a7ddb5f6.png)
![1618497391](https://user-images.githubusercontent.com/11150153/114887464-b4100900-9dc5-11eb-9fd1-36348a55ab3c.png)

this will only manifest after reading the sequences from the file I shared, not before, but again, tmux 3.2c does not  have this issue, should I open an issue for this?

**Comment by nicm** [comment]
tmux 3.1 does not support setting the palette in the way you are doing in that file. Yes please open an issue with logs.

**Comment by Flexximilian** [comment]
Hi! Thanks for gifting us with tmux, much appreciated, indeed.

There appears to be a regression in tmux 3.2 (on linux) where 

`tmux new-session -n test \; split-window -dvb -l1` (with no previous session running)

used to work as expected when called from the (bash) command line: it would open a new session with two panes -- a 1-line pane ontop of another pane occupying the rest of the window. In 3.2 though, this does open two panes, but the top pane is larger than expected (>1 line if the terminal is large enough). 

Interestingly, when calling `tmux` from the command line, followed by `tmux split-window -dvb -l1` inside, it works, you get a single-line pane on the top. I've tried arguments 1, 2, 3, 4 for `-l`. Works as expected inside, but when given as above for a new session, the top window grows with increasing numbers, but it almost looks like the top pane gets 50% _minus_ the given number of lines.

I've tested this on current KDE Konsole ($TERM=xterm-256colors), xterm ($TERM=xterm) and the Linux console ($TERM=linux); inside $TERM=screen for all three cases. Would you like me to open an issue for this?

**Comment by Federico-Ciuffardi** [comment]
All my user-keys bindings stopped working on the latest update, I just cloned the repo and installed tmux from source which seems to contain the following fix :
> Actually please try this instead: https://gist.githubusercontent.com/nicm/be4faa3ff53884384695fc271b3428fe/raw/6cfa7f662a61eae060e8878c6230dfceb6879c70/gistfile1.txt

But the user-keys bindings still not working. Some examples of the user-keys bindings are:
```
set -s user-keys[0] "\033[9;5u" # C-Tab
# preview and switch session
bind-key -n User0 run "$TMUX_BIN/switch_session '#{session_id}'" 

set -s user-keys[1] "\033[84;6u" # C-S-T
bind-key -n User1 new-window

set -s user-keys[2] "\033[13;6u" # C-S-Enter
bind-key -n User2 new-window

set -s user-keys[3] "\033[75;6u" # C-S-J
bind-key -n User3 next-window
```
Please let me know if I missed something, or if you need any more information.

**Comment by nicm** [comment]
@Federico-Ciuffardi I meant to make user-keys checked before extended-keys but I forgot, so you will need to bind the keys as `C-Tab`, `C-S-t` etc instead. If that does not work please open an issue.

**Comment by nicm** [comment]
@Flexximilian give `-x-` and `-y-` to `new-session`.

**Comment by amukherjee-fincad** [comment]
@nicm @ChrispyChris re: the search counter request, I gave it further thought and I think I have a scheme where we can accurately maintain a current match count (if there is no timeout) without doing any additional searches besides the very first one. It will also not complicate the existing code too much, only adding a little bit for the counter and a bit of bookkeeping. I'll follow it up in a separate PR when I'm ready so that you can see if you like it.

**Comment by topcat001** [comment]
@nicm @ChrispyChris re: the search counter request, I gave it further thought and I think I have a scheme where we can accurately maintain a current match count (if there is no timeout) without doing any additional searches besides the very first one. It will also not complicate the existing code too much, only adding a little bit for the counter and a bit of bookkeeping. I'll follow it up in a separate PR when I'm ready so that you can see if you like it.

**Comment by nicm** [comment]
Great stuff, thanks.

**Comment by Federico-Ciuffardi** [comment]
@nicm it works for C-Tab and C-S-t, but:
* C-S-J binding triggers when C-S-K is pressed and vice versa
* C-S-Enter does not work
* Can't find how to bind to some keys combinations, like for example: C--, C-`, C-S-Backspace

**Comment by nicm** [comment]
Open an issue with logs from pressing these keys please.

**Comment by Federico-Ciuffardi** [comment]
Ok, i just opened an issue!

But will the user-keys bindings work on the future?

**Comment by Memphizzz** [comment]
Are there any plans to create an official launchpad.net archive so we can add a ppa on ubuntu systems to get the latest version and not have to rely on the version in the ubuntu repository?

**Comment by jorgeOmurillo** [comment]
On Mac Big Sur, v3.1 automatically loaded my config file in `~/.config/tmux/tmux.conf`, but v3.2 doesn't and I need to manually source the configuration to get it working `tmux source-file /Users/myUser/.config/tmux/tmux.conf`. Any help?

**Comment by krader1961** [comment]
@jorgeOmurillo: Tmux has never automatically loaded *~/.config/tmux/tmux.conf*. It is likely you made a symlink from *~/.tmux.conf* to that path; or, less likely, you built a custom tmux binary. For example, I like to keep my "dot files" in a shared filesystem:

```
> ls -ld .tmux.conf
lrwxr-xr-x 1 krader staff 26 May  1  2014 .tmux.conf -> Dropbox/dotfiles/tmux.conf
```

**Comment by gpanders** [comment]
> @jorgeOmurillo: Tmux has never automatically loaded _~/.config/tmux/tmux.conf_. It is likely you made a symlink from _~/.tmux.conf_ to that path; or, less likely, you built a custom tmux binary.

This is incorrect. In tmux 3.1, support was added for `~/.config/tmux/tmux.conf`. In 3.2 support was also added for `$XDG_CONFIG_HOME/tmux/tmux.conf`.

@jorgeOmurillo check to see if your `$XDG_CONFIG_HOME` variable is set or if maybe you somehow have a `.tmux.conf` file in your home directory that is being sourced instead.

**Comment by krader1961** [comment]
@gpanders: You're correct. I didn't notice the added support for the XDG standard because the project owner has adamantly rejected adding support for XDG for so many years I just assumed it still hadn't happened :smile:. It looks like it was added on 2019-12-10 by git commit 15d7e564d. Looking at that change it's unclear if the feature was introduced in the 3.0 or a later release.

**Comment by systemmonkey42** [comment]
> @jorgeOmurillo check to see if your `$XDG_CONFIG_HOME` variable is set or if maybe you somehow have a `.tmux.conf` file in your home directory that is being sourced instead.

Be aware that all matching configs (~/.tmux.conf, $XDG_CONFIG_HOME/tmux/tmux.conf and ~/.config/tmux/tmux.conf) will be loaded (probably in that order).

I deploy my "default" config in ~/.config/tmux/tmux.conf which is present on all systems I use, and customise with local ~/.tmux.conf files.

An error in one does not appear to stop the loading of subsequent config files, though this might be new behaviour.

**Comment by jorgeOmurillo** [comment]
Thanks @gpanders, that was it. My `XDG_CONFIG_HOME` was empty.

**Comment by bpsib** [comment]
> found a consistent way to reproduce, download the following into a file called `sequences`: https://u.teknik.io/XXOWm
> 
> I tried pasting the contents here but it lost information due to it using special characters or something, posting the content just for reference but for it to work you're gonna need to use the file provided

@deviantfero Can you upload that file again? I get 404 not found when I click the link.

**Comment by bpsib** [comment]
I have noticed that `choose-client` (`Prefix`+`Shift`+`D`) no longer shows the window sizes of each client. I added `bind -T prefix D choose-client -Z -F "#{t:client_activity}: session #{client_session}: #{client_width}x#{client_height}"` to my config to get it back showing the sizes.

**Comment by cmcginty** [comment]
This macro I had in my config file no longer parses in v3.2? Should it still be working?
```
TMUX_OFF_MACRO={
  set prefix None
  set key-table off
  set -F status-style '#{@BAR_OFF_STYLE}'
  set -F window-status-current-style '#{@BAR_OFF_STYLE}'
  set -F window-status-current-format '#{@WIN_STATUS_OFF}'
  set -F status-left '#{@STATUS_OFFL}'
  set -F status-right '#{@STATUS_OFFR}'
  set status on
  if -F '#{pane_in_mode}' 'send-keys -X cancel'
  refresh-client -S
}
```

**Comment by Perdu** [comment]
Hello,
I used to have this config working to switch between tmux sessions, while defining different variables for these sessions (notably bar color):
`bind-key -n M-S-Up switch-client -n \; source-file ~/.tmux.conf.sessions`
In other words: switch session, then reload config for that session (in ~/.tmux.conf.sessions, I have tests on #{session_id} and #{session_name}).
But since 3.2, the logic seems to have changed, and source-file will be called with pre-switch variables, i.e. variables from the previous session.
Is that normal or a bug?
Is there a workaround?

**Comment by vbauerster** [comment]
@nicm
Extended keys don't work in v3.2. I tested with iTerm2 both stable and beta with following tmux config:

```
set-option -s extended-keys on
set-option -as terminal-features 'xterm*:extkeys'
```

So in v3.2 following test with `C-1` doesn't work (it produces just 1 instead of `^[[49;5u`), but it worked in v3.2-rc4.

```
$ printf '\033[>4;1m'
$ cat
^[[49;5u
```

Is there any extra config needed to enable extended keys?

**Comment by vbauerster** [comment]
Is `extended-keys` mode handling `C-m`, `C-j` and `C-i`?

**Comment by smhc** [comment]
> * Add a -w flag to set- and load-buffer to send to clipboard using OSC 52.

Is there a way to make this the default behaviour for load-buffer?

neovim uses load-buffer to integrate with tmux during copy/paste but does not pass this switch. It can be configured to do this by overriding neovim config and adding the -w switch:

https://gist.github.com/smhc/7b873a553d75af7142f38b7993f5f0ec

However ideally tmux should control whether it uses osc52 or not rather than neovim. neovim should ideally be tmux version agnostic and simply inherit the tmux configured behaviour.

**Comment by nicm** [comment]
No I do not want this to be the default behaviour.

**Comment by smhc** [comment]
@nicm 

I was asking whether it was possible to configure it to be the default behaviour of running load-buffer, rather than the default behaviour for everyone. i.e 

set clipboard on-with-osc

would result in 'load-buffer' running with an implied '-w' switch.

**Comment by nicm** [comment]
You should be able to use `command-alias` for that, but how is configuring it in tmux any different from configuring it in vim?

**Comment by smhc** [comment]
Not a lot different. In theory there could be multiple applications using load-buffer and you want a single place to configure whether OSC-52 is used or not. There's an argument that some users would want tmux config to control whether it occurs or not.

It's also not ideal to override the neovim config that supports tmux out of the box. I suppose neovim could add a 'g:tmux_clipboard_osc52=1' flag which adds this -w flag instead of having users duplicate the core clipboard config to add the extra flag. If tmux won't add the '-w by default' behaviour this might be an enhancement I raise with neovim.

**Comment by MiralDesai** [comment]
Question, am I right in thinking a bunch of keyboard shortcuts no longer work in this version? I'm using mac os big sur, I have one laptop with tmux 3.2 running and another with the same OS and tmux 3.1c running. Both using the same `.tmux.conf`.

On 3.1c I can switch between pages with `option` + `cmd` + `arrow keys`.

I can also resize panes by first hitting my modifier key(default `C^b` in my case) then doing `option` + `arrow keys`

Both of these no longer work in 3.2. Just curious if there's anything I need to do to get these working again, from what I understand these were default commands that work without requiring anything in `.tmux.conf`

**Comment by nicm** [comment]
Do they work if you build tmux from master?

**Comment by MiralDesai** [comment]
Can confirm is does work on master.

**Comment by rfrancoise** [comment]
Hi @nicm, any update on a 3.2a release?

**Comment by systemmonkey42** [comment]
Hey,

Has 'tmux popup' changed significantly?

I a recent (next-3.3) build (maybe 2 weeks?), running `tmux popup ls` gives me a rectangle in the centre of the window containing the output of the shell command `ls`

Have built from master today, `tmux popup ls` gives me an empty popup.  When I close the popup,  I get the output of "tmux ls" in my active pane, which I then have to exit out of.

Is this new functionality? It seems broken to me.
Thanks

**Comment by nicm** [comment]
@rfrancoise no I have not had time yet.

**Comment by invsblduck** [comment]
@nicm 
> It was removed because it didn't work in 3.1, it is much faster not to try and keep track of it, and neither vi or emacs show a match count. I don't think we need it.

we don't _really_ need it (especially if it affects performance or isn't O(1) time), but i find the **total** number of matches to be an extremely helpful indicator&#8212;it immediately orients me and sets my expectations as to the usefulness of my search. i don't really care whether i'm on match 7 of 18, but i care to know there are 18 results and not 18,000 when i start pressing `n`. &nbsp; :-)

**Comment by topcat001** [comment]
The total number of matches is already displayed currently (if it does not time out). Does that address your needs? I might have missed something here.

I am planning to bring back the current match counter using a more efficient method which will not search the whole thing every time; just didn't get to it yet. Hopefully soon and then we can see how well it works.

**Comment by ltning** [comment]
After upgrading from 3.1c to 3.2, mouse handling seems broken in my use case, which involves nesting (two levels) of tmux.
Mouse events (scrolling, in particular) are passed to the outermost tmux only, whereas previously they would be passed to whichever pane had focus.
I've tried to read the changelog (to the point that my eyes are now running ;), but apart from the "it's a lot", nothing obvious stands out.

Halp?

**Comment by nicm** [comment]
If you do `printf "\033[?1000h\033[?1006h";cat` in a shell inside the outer tmux and click the mouse, do you see mouse sequences printed?

**Comment by ltning** [comment]
In outer tmux session, yes:
```
$ printf "\033[?1000h\033[?1006h";cat
^[[<0;105;25M^[[<0;105;25m
```

**Comment by nicm** [comment]
What does `tmux show -g mouse` show in the inner tmux?

**Comment by ltning** [comment]
```
$ tmux show -g mouse
mouse on
```
in both inner and outer tmux.

**Comment by nicm** [comment]
Open a new issue please with logs from the inner tmux.

**Comment by ltning** [comment]
Ok - how do I collect relevant logs?

**Comment by nicm** [comment]
https://github.com/tmux/tmux/blob/master/.github/CONTRIBUTING.md#how-do-i-get-logs-from-tmux

**Comment by ltning** [comment]
Thanks, and sorry for being dense and not looking that up myself. Issue raised, #2721 .

**Comment by invsblduck** [comment]
> The total number of matches is already displayed currently (if it does not time out). Does that address your needs? I might have missed something here.

@topcat001 Doh. :man_facepalming: Thank you. That's my bad since I haven't actually *used* 3.2 yet (I will unpin `3.1_c` when I feel ready). It wasn't clear from the comments whether there still was a total-matches number anymore, so I chimed in about the use-case of total vs. current. THANK YOU :+1:

**Comment by systemmonkey42** [comment]
Hi,

What is the most effective way to get #{...} expanded in the display-popup command?

**Comment by nicm** [comment]
You can use run -C which will expand formats in its argument.
